### PR TITLE
[Different approach] Fix entity destruction in unloaded chunks

### LINKED ISSUE #2731: `/butcher` command reports too many mobs
The butcher command seems to be lying about the number of mobs killed. This is either some internal issue or a plugin issue.

### LINKED ISSUE #2973: Investigate entity removal in unloaded chunks
Removing entities in unloaded chunks was the prime cause of world instability. It delayed the removal until the chunk loads, often causing m_Entities corruption in the process. #2941 resolves this by adding a function called  [cChunk::SafeRemoveEntity](https://github.com/cuberite/cuberite/commit/07b7fd4ad3fef714399009efd5d4a903a561a6f2#diff-af2ddf3aaa057bf2bbd7dbfe509c9918R265). We should investigate if more code paths lead to entity destruction in unloaded chunks, and we should make those code paths use some variant / modification of [cChunk::SafeRemoveEntity](https://github.com/cuberite/cuberite/commit/07b7fd4ad3fef714399009efd5d4a903a561a6f2#diff-af2ddf3aaa057bf2bbd7dbfe509c9918R265). #2731 is caused by such a code path.

### PR DESCRIPTION
**Pull request by LogicParrot**
Work in progress. This is the competitor PR to #2980

Opinions?

Closes #2731
Closes #2973
May close or alleviate #277

### CODE REVIEW

File: src/Chunk.cpp [hunk 1]
```diff
@@ -1937,6 +1940,24 @@ void cChunk::SafeRemoveEntity(cEntity * a_Entity)
...
+	cEntity * CurrentTickingEntity = *m_CurrentEntityIterator;
+
+	if (CurrentTickingEntity != a_Entity)
+	{
+		LOGD("Destroying entity #%i (%s) (scheduled)", a_Entity->GetUniqueID(), a_Entity->GetClass());
+		// If we're not in a tick, just destroy us.
+		m_Entities.remove(a_Entity);
+		delete a_Entity;
+	}
+	// Else, the entity be removed in cChunk::tick() @ "if ((*itr)->IsDestroyed())"
```
Comment (LogicParrot) [inline src/Chunk.cpp]: will be removed

File: src/Entities/Entity.cpp [hunk 2]
```diff
@@ -40,7 +40,6 @@ cEntity::cEntity(eEntityType a_EntityType, double a_X, double a_Y, double a_Z, d
 	m_AirDrag(0.02f),
 	m_LastPosition(a_X, a_Y, a_Z),
 	m_IsInitialized(false),
-	m_WorldTravellingFrom(nullptr),
```
Comment (LogicParrot) [inline src/Entities/Entity.cpp]: This should have stayed here.

File: src/World.cpp [hunk 3]
```diff
@@ -1088,9 +1088,13 @@ void cWorld::TickMobs(std::chrono::milliseconds a_Dt)
 
 	// move close mobs
 	cMobProximityCounter::sIterablePair allCloseEnoughToMoveMobs = MobCensus.GetProximityCounter().getMobWithinThosesDistances(-1, 64 * 16);// MG TODO : deal with this magic number (the 16 is the size of a block)
-	for (cMobProximityCounter::tDistanceToMonster::const_iterator itr = allCloseEnoughToMoveMobs.m_Begin; itr != allCloseEnoughToMoveMobs.m_End; ++itr)
+	for (cMobProximityCounter::tDistanceToMonster::const_iterator itr = allCloseEnoughToMoveMobs.m_Begin; itr != allCloseEnoughToMoveMobs.m_End;)
 	{
+		cMobProximityCounter::tDistanceToMonster::const_iterator Next = ++itr;
+		--itr;
+		// Ticking might destroy the monster if it calls destroy() (it would be deleted in cChunk::SafeDestroyEntity), so we preserve the NEXT itr.
 		itr->second.m_Monster.Tick(a_Dt, itr->second.m_Chunk);
```
Comment (LogicParrot) [inline src/World.cpp]: This crashes sometimes. (Use after free, probably).

File: src/Chunk.cpp [hunk 4]
```diff
@@ -1937,6 +1940,24 @@ void cChunk::SafeRemoveEntity(cEntity * a_Entity)
 
 
 
+void cChunk::SafeDestroyEntity(cEntity * a_Entity)
+{
+	cEntity * CurrentTickingEntity = *m_CurrentEntityIterator;
+
+	if (CurrentTickingEntity != a_Entity)
+	{
+		LOGD("Destroying entity #%i (%s) (scheduled)", a_Entity->GetUniqueID(), a_Entity->GetClass());
```
Comment (LogicParrot) [inline src/Chunk.cpp]: immediate, not scheduled.

File: src/Chunk.cpp [hunk 5]
```diff
@@ -621,21 +620,23 @@ void cChunk::Tick(std::chrono::milliseconds a_Dt)
 
 	for (cEntityList::iterator itr = m_Entities.begin(); itr != m_Entities.end();)
 	{
-		if (!((*itr)->IsMob()))  // Mobs are ticked inside cWorld::TickMobs() (as we don't have to tick them if they are far away from players)
+		cEntityList::iterator Next = ++itr;  // BeginTick may destroy the entity, soo keep the next iterator.
```
Comment (LogicParrot) [inline src/Chunk.cpp]: so

File: src/Chunk.h [hunk 6]
```diff
@@ -505,11 +505,10 @@ class cChunk :
 
 	/** If the chunk fails to load, should it be queued in the generator or reset back to invalid? */
 	bool m_ShouldGenerateIfLoadFailed;
-	bool m_IsInTick;       // True if the chunk is executing the tick() method.
-	bool m_IsLightValid;   // True if the blocklight and skylight are calculated
-	bool m_IsDirty;        // True if the chunk has changed since it was last saved
-	bool m_IsSaving;       // True if the chunk is being saved
-	bool m_HasLoadFailed;  // True if chunk failed to load and hasn't been generated yet since then
+	bool m_IsLightValid;               // True if the blocklight and skylight are calculated
```
Comment (LogicParrot) [inline src/Chunk.h]: remove extra spaces

File: src/Entities/Entity.h [hunk 7]
```diff
@@ -336,6 +336,8 @@ class cEntity
 
 	// tolua_end
 
+	bool BeginTick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk);
```
Comment (LogicParrot) [inline src/Entities/Entity.h]: Document return vals

File: src/Entities/Entity.h [hunk 8]
```diff
@@ -481,7 +477,14 @@ class cEntity
 	/** Sets the internal world pointer to a new cWorld, doesn't update anything else. */
 	void SetWorld(cWorld * a_World) { m_World = a_World; }
 
+	/** Sets the chunk which is owning us. Should only be called by cChunk::addEntity. */
+	void SetParentChunk(cChunk * a_Chunk);
+
+	/** Sets whether we're removed or not. Should only be called by cChunk::addEntity or cEntity::BeginTick */
```
Comment (LogicParrot) [inline src/Entities/Entity.h]: Outdated documentation

File: src/World.cpp [hunk 9]
```diff
@@ -1088,9 +1088,16 @@ void cWorld::TickMobs(std::chrono::milliseconds a_Dt)
...
-	for (cMobProximityCounter::tDistanceToMonster::const_iterator itr = allCloseEnoughToMoveMobs.m_Begin; itr != allCloseEnoughToMoveMobs.m_End; ++itr)
+	for (cMobProximityCounter::tDistanceToMonster::const_iterator itr = allCloseEnoughToMoveMobs.m_Begin; itr != allCloseEnoughToMoveMobs.m_End;)
 	{
-		itr->second.m_Monster.Tick(a_Dt, itr->second.m_Chunk);
+		cMobProximityCounter::tDistanceToMonster::const_iterator Next = ++itr;  // Ticking might destroy the monster, so we preserve the next iterator.
+		--itr;
+		bool Destroyed = itr->second.m_Monster.BeginTick(a_Dt, itr->second.m_Chunk);
+		if (Destroyed)
+		{
+			itr->second.m_Chunk.MarkDirty();
```
Comment (LogicParrot) [inline src/World.cpp]: No need for this. This will happen already: cEntity::beginTick > cChunk::DestroyEntity

**Comment by tigerw** [comment]
Hmm. Way too many new variables added. It's getting ever more difficult to follow control flow in the entity management system.

The correct way to do this is to have Destroy queue a task onto the entity's world, which calls a function in cChunk that removes the entity.

**Comment by LogicParrot** [comment]
I'll see what I can do about the new variables. I might be able to simplify the flow.

I oppose a queue. I want destruction to occur ASAP. (Hence destruction is called in BeginTick).

**Comment by LogicParrot** [comment]
@tigerw Is there any reason to use a queue? I don't see any problem with destroying entities right away. Assuming this code is simplified, would a queue still be beneficial?

**Comment by tigerw** [comment]
The thing is, not using a queue means everything turns into spaghetti. BeginTick? IsInTick? It's not immediately obvious why so many single threaded synchronisation variables are needed.

It's a situation of:
1. Destroy()
2. cChunk performs gymnastics to try to modify m_Entities whilst in Tick(), or else directly remove it whilst out of Tick()
3. Entity removed in same tick

versus

1. Destroy()
2. Task queued onto world, cWorld::cChunkMap::Tick() returns
3. ProcessQueuedTasks runs immediately afterwards, and removes the entity
4. Still removed in same tick

**Comment by tigerw** [comment]
How urgent is this fix? If not very, I'm doing a threading rework which can incorporate this.

**Comment by LogicParrot** [comment]
@tigerw Currently many entities are left in the limbo instead of getting truly destroyed. I don't know how urgent are the side effects of that but this is supposed to fix it.

**Comment by LogicParrot** [comment]
@tigerw Okay, your idea of a world-wide queue sounds reasonable. But working code is better than a bug. So I propose we merge this, and you're free to further improve it by implementing the queue.

Broken Server < this PR < Queue

**Comment by LogicParrot** [comment]
Unless you plan on doing the queue real soon.

**Comment by LogicParrot** [comment]
@tigerw How does the world interact with plugins time-wise? Suppose a plugin calls an entity's Destroy, would the world still tick and process the queue?

**Comment by tigerw** [comment]
The world tick thread never stops ticking.

**Comment by LogicParrot** [comment]
Okay. I think I've come up with a very simple queue implementation. PR soon.

**Comment by LogicParrot** [comment]
Closing in favor of #2986

**Pull request closed without merging**